### ISSUE TITLE
`cargo dylint my_lints` produces `Error: expected more input` while parsing crate versions after updating `rustc`

### ISSUE THREAD

**Issue by PatchMixolydic** [trailofbits/dylint#19]
It seems like the error happens around here:

https://github.com/trailofbits/dylint/blob/b9002a5e8b999601326c28870b4775c5f3dfe884/dylint/src/driver_builder.rs#L107-L108

`rustc -Vv`:
```
rustc 1.54.0-nightly (5dc8789e3 2021-05-21)
binary: rustc
commit-hash: 5dc8789e300930751a78996da0fa906be5a344a2
commit-date: 2021-05-21
host: x86_64-unknown-linux-gnu
release: 1.54.0-nightly
LLVM version: 12.0.1
```

Lint library's Cargo.toml:
```toml
[package]
name = "patchmixolint"
version = "0.1.0"
authors = ["Ruby Lazuli"]
description = "Set of personal lints"
edition = "2018"
publish = false

[lib]
crate-type = ["cdylib"]

[dependencies]
clippy_utils = { git = "https://github.com/rust-lang/rust-clippy", rev = "ae0d4da764fd751494fb270fd04c2c686eac1302"}
dylint_linting = "0.1.0"
if_chain = "1.0.1"

[dev-dependencies]
dylint_testing = "0.1.0"

[package.metadata.rust-analyzer]
rustc_private = true
```

**Comment by PatchMixolydic** [comment]
Deleting and rebuilding the driver seems to have fixed this issue, so this may have to do with using an old driver on a newer `rustc`. For clarity's sake, it was an upgrade from 1.54.0-nightly 2020-05-10 to 1.54.0-nightly 2020-05-21.

**Comment by smoelius** [comment]
Thanks a lot for the report, @PatchMixolydic. Could you please tell me what version of `cargo-dylint` you're using?
```sh
cargo dylint --version
```
I'm not 100% certain, but I think I may have run into this a few days ago and fixed it with: https://github.com/trailofbits/dylint/commit/c28639eecefb88c2d95e67527239600867b04757

So, this problem may be fixed in version `0.1.2`.

Either way, it sounds like you're not experiencing the problem now?

**Comment by PatchMixolydic** [comment]
I'm on `0.1.2`. I have not experienced the problem since deleting and rebuilding the driver.

**Comment by smoelius** [comment]
>  If I remember correctly, I tried updating both dylint-link and cargo-dylint to no avail.

Meaning you experienced the problem, updated the two binaries, *still* experienced the problem, and it finally went away after you deleted and rebuilt the driver?

That would be conducive with my experience before https://github.com/trailofbits/dylint/commit/c28639eecefb88c2d95e67527239600867b04757. Failing to install the required `rustc` components was somehow causing corrupt drivers to be produced. I also have not observed the issue since updating.

Is it okay if I close this issue? Needless to say, you're welcome to reopen if the problem arises again.

**Comment by PatchMixolydic** [comment]
> Meaning you experienced the problem, updated the two binaries, still experienced the problem, and it finally went away after you deleted and rebuilt the driver?

That's correct.

> Is it okay if I close this issue? Needless to say, you're welcome to reopen if the problem arises again.

Sure :+1: I've [put my library on GitHub](https://github.com/PatchMixolydic/patchmixolint), so feel free to check for anything strange. One suspect that's come to mind is that I'm using `channel = "nightly"` in `rust-toolchain` to track the currently installed nightly version, which may be a bad idea:

https://github.com/PatchMixolydic/patchmixolint/blob/5c865f89de49e15216b1ef96575e88af008e715b/rust-toolchain#L2

**Comment by smoelius** [comment]
> Sure 👍 I've put my library on GitHub, so feel free to check for anything strange.

These look fantastic! I especially like `impl_eq_for_float`.

> A notable thing that's just come to mind is that I'm using channel = "nightly" in rust-toolchain to track the currently installed nightly version, which may be a bad idea.

This isn't terrible, but I think what may eventually happen is that one of the lints or `clippy_utils` will fail to compile. Of course, you could wait for this to happen, and then roll back to the last compiler version where things worked.

Thanks very much for sharing and thanks for this report! Please let me know if the problem arises again.

**Comment by rukai** [comment]
I am getting the same issue:
```
Stack backtrace:
   0: anyhow::error::<impl core::convert::From<E> for anyhow::Error>::from
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/anyhow-1.0.42/src/error.rs:519:25
   1: <core::result::Result<T,F> as core::ops::try_trait::FromResidual<core::result::Result<core::convert::Infallible,E>>>::from_residual
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/core/src/result.rs:1653:27
   2: dylint::driver_builder::is_outdated
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/dylint-1.0.0/src/driver_builder.rs:108:25
   3: dylint::driver_builder::get
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/dylint-1.0.0/src/driver_builder.rs:76:28
   4: dylint::check
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/dylint-1.0.0/src/lib.rs:407:22
   5: dylint::run_with_name_toolchain_map
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/dylint-1.0.0/src/lib.rs:170:9
   6: dylint::run
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/dylint-1.0.0/src/lib.rs:64:5
   7: cargo_dylint::cargo_dylint
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-dylint-1.0.0/src/main.rs:174:37
   8: cargo_dylint::main
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-dylint-1.0.0/src/main.rs:160:18
   9: core::ops::function::FnOnce::call_once
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/core/src/ops/function.rs:227:5
  10: std::sys_common::backtrace::__rust_begin_short_backtrace
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/sys_common/backtrace.rs:125:18
  11: std::rt::lang_start::{{closure}}
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/rt.rs:63:18
  12: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/core/src/ops/function.rs:259:13
  13: std::panicking::try::do_call
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/panicking.rs:401:40
  14: std::panicking::try
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/panicking.rs:365:19
  15: std::panic::catch_unwind
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/panic.rs:434:14
  16: std::rt::lang_start_internal::{{closure}}
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/rt.rs:45:48
  17: std::panicking::try::do_call
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/panicking.rs:401:40
  18: std::panicking::try
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/panicking.rs:365:19
  19: std::panic::catch_unwind
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/panic.rs:434:14
  20: std::rt::lang_start_internal
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/rt.rs:45:20
  21: std::rt::lang_start
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/rt.rs:62:5
  22: main
  23: __libc_start_main
  24: _start
```
I am running the lint like this:
`cargo dylint --path /home/rukai2/Projects/Crates/dylint/examples/allow_clippy/target/release/liballow_clippy@nightly-2021-07-01-x86_64-unknown-linux-gnu.so`


It seems to occur when run on a complicated project https://github.com/embassy-rs/embassy (Needs a recent nightly so I had to upgrade the lint to run on nightly-2021-07-04 but not when run on a simple hello world project.

When I run the lint on the hello world project I also seem to be getting no output even when the lint should fail.

**Comment by smoelius** [comment]
Hi, @rukai. That error means Dylint cannot determine the version of the driver it needs to run.

Please try `rm -rf ~/.dylint_drivers` and then re-run your command.

If that does not work, could you please tell me the output of the following commands?
```
cargo dylint --version
ls ~/.dylint_drivers
file ~/.dylint_drivers/*/*
```

**Comment by rukai** [comment]
I deleted `~/.dylint_drivers` and now im getting this:
```
rukai@memes foo$ ./lint.sh 
    Updating crates.io index
error: failed to get `dylint_driver` as a dependency of package `dylint_driver-nightly-2021-07-01-x86_64-unknown-linux-gnu v0.1.0 (/tmp/.tmpUoPvYq)`

Caused by:
  failed to load source for dependency `dylint_driver`

Caused by:
  Unable to update /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/driver

Caused by:
  failed to read `/home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/driver/Cargo.toml`

Caused by:
  No such file or directory (os error 2)
If you don't see a backtrace below, it could be because `cargo-dylint` wasn't built with a nightly compiler.
Error: command failed: "cargo" "build"

Stack backtrace:
   0: anyhow::private::new_adhoc
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/anyhow-1.0.42/src/lib.rs:632:36
   1: dylint_internal::command::Command::success
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/dylint_internal-1.0.0/src/command.rs:72:9
   2: dylint::driver_builder::build
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/dylint-1.0.0/src/driver_builder.rs:158:5
   3: dylint::driver_builder::get
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/dylint-1.0.0/src/driver_builder.rs:77:9
   4: dylint::check
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/dylint-1.0.0/src/lib.rs:407:22
   5: dylint::run_with_name_toolchain_map
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/dylint-1.0.0/src/lib.rs:170:9
   6: dylint::run
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/dylint-1.0.0/src/lib.rs:64:5
   7: cargo_dylint::cargo_dylint
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-dylint-1.0.0/src/main.rs:174:37
   8: cargo_dylint::main
             at /home/rukai/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-dylint-1.0.0/src/main.rs:160:18
   9: core::ops::function::FnOnce::call_once
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/core/src/ops/function.rs:227:5
  10: std::sys_common::backtrace::__rust_begin_short_backtrace
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/sys_common/backtrace.rs:125:18
  11: std::rt::lang_start::{{closure}}
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/rt.rs:63:18
  12: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/core/src/ops/function.rs:259:13
  13: std::panicking::try::do_call
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/panicking.rs:401:40
  14: std::panicking::try
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/panicking.rs:365:19
  15: std::panic::catch_unwind
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/panic.rs:434:14
  16: std::rt::lang_start_internal::{{closure}}
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/rt.rs:45:48
  17: std::panicking::try::do_call
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/panicking.rs:401:40
  18: std::panicking::try
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/panicking.rs:365:19
  19: std::panic::catch_unwind
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/panic.rs:434:14
  20: std::rt::lang_start_internal
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/rt.rs:45:20
  21: std::rt::lang_start
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/rt.rs:62:5
  22: main
  23: __libc_start_main
  24: _start
```

**Comment by smoelius** [comment]
Admittedly, I am a bit confused. That error suggests that Dylint is trying to build the driver using a local copy of the `dylint_driver` package as opposed to one obtained from `crates.io`.

Here is the relevant code: https://github.com/trailofbits/dylint/blob/c6138812e54460c3d43750e8b378ff2568c4cbea/dylint/src/driver_builder.rs#L129-L137

As you can see from the use of `debug_assertions`, only a debug build should try to reference the `dylint_driver` package in that way.

But based on the invocation you gave earlier, I would assume that you `cargo install`ed `cargo-dylint`. (Is this correct?)

And `cargo install` should produce a release build, not a debug build.

You seemed to have built a driver earlier (albeit a potentially corrupt one). Can you think of what might have changed since that driver was built?

**Comment by rukai** [comment]
I did `cargo +nightly install cargo-dylint --force --debug` because I wanted to be able to get a proper stacktrace from the panic.
Will using a debug build cause issues?

**Comment by smoelius** [comment]
Short answer: yes.

Long answer: What the code I linked above tries to achieve is the following.
* When building a driver from a `cargo install`ed `dylint` package, obtain the `dylint_driver` package from `crates.io`.
* When building a driver from a GitHub-cloned `dylint` package, obtain the `dylint_driver` package from the same clone of the repository.

To achieve this, I need a way to determine whether the `dylint` package was `cargo install`ed or not, and I am currently using release vs. debug as an approximation. I am open to alternatives if better ones exist. (TBH, I didn't know `cargo install` had a `--debug` flag.)

Could you try without the `--debug` flag and see if you get different results?

**Comment by rukai** [comment]
Ah ok.
I think inspecting the path of the dylint binary should give more accurate results for that.

I reinstalled with --release.
It then built once but not sure if it worked properly as I dont think the lint should have failed on this project.
And then when I rerun I get "expected more input" error.
```
rukai@memes embassy$ cargo dylint --path /home/rukai2/Projects/Crates/deny_core_fmt/deny_core_fmt_dylint/target/release/libdeny_core_fmt@nightly-2021-07-01-x86_64-unknown-linux-gnu.so
    Updating crates.io index
   Compiling libc v0.2.98
   Compiling pkg-config v0.3.19
   Compiling autocfg v1.0.1
   Compiling tinyvec_macros v0.1.0
   Compiling matches v0.1.8
   Compiling log v0.4.14
   Compiling cfg-if v1.0.0
   Compiling percent-encoding v2.1.0
   Compiling memchr v2.4.0
   Compiling bitflags v1.2.1
   Compiling ryu v1.0.5
   Compiling serde v1.0.126
   Compiling anyhow v1.0.42
   Compiling serde_json v1.0.64
   Compiling openssl-probe v0.1.4
   Compiling regex-syntax v0.6.25
   Compiling itoa v0.4.7
   Compiling humantime v2.1.0
   Compiling termcolor v1.1.2
   Compiling tinyvec v1.2.0
   Compiling libloading v0.7.0
   Compiling unicode-bidi v0.3.5
   Compiling form_urlencoded v1.0.1
   Compiling aho-corasick v0.7.18
   Compiling jobserver v0.1.22
   Compiling atty v0.2.14
   Compiling unicode-normalization v0.1.19
   Compiling cc v1.0.68
   Compiling regex v1.5.4
   Compiling idna v0.2.3
   Compiling openssl-sys v0.9.65
   Compiling libz-sys v1.1.3
   Compiling libssh2-sys v0.2.21
   Compiling libgit2-sys v0.12.21+1.1.0
   Compiling env_logger v0.8.4
   Compiling url v2.2.2
   Compiling git2 v0.13.20
   Compiling dylint_internal v1.0.0
   Compiling dylint_driver v1.0.0
   Compiling dylint_driver-nightly-2021-07-01-x86_64-unknown-linux-gnu v0.1.0 (/tmp/.tmpgr4D1D)
    Finished dev [unoptimized + debuginfo] target(s) in 8.89s
   Compiling semver-parser v0.7.0
   Compiling proc-macro2 v1.0.26
   Compiling unicode-xid v0.2.2
   Compiling syn v1.0.72
    Checking nb v1.0.0
   Compiling cortex-m v0.7.2
   Compiling futures-core v0.3.15
    Checking void v1.0.2
   Compiling futures-task v0.3.15
   Compiling autocfg v1.0.1
   Compiling ident_case v1.0.1
   Compiling strsim v0.9.3
   Compiling fnv v1.0.7
    Checking vcell v0.1.3
    Checking futures-sink v0.3.15
   Compiling futures-channel v0.3.15
    Checking pin-project-lite v0.2.6
   Compiling atomic-polyfill v0.1.2
    Checking bitfield v0.13.2
   Compiling critical-section v0.2.1
    Checking pin-utils v0.1.0
    Checking bare-metal v1.0.0
    Checking futures-io v0.3.15
    Checking cfg-if v1.0.0
    Checking nb v0.1.3
    Checking volatile-register v0.2.0
   Compiling semver v0.9.0
    Checking embedded-hal v0.2.5
   Compiling embassy v0.1.0 (/home/rukai2/Projects/Crates/embassy/embassy)
   Compiling futures-util v0.3.15
    Checking embassy-traits v0.1.0 (/home/rukai2/Projects/Crates/embassy/embassy-traits)
   Compiling rustc_version v0.2.3
   Compiling bare-metal v0.2.5
   Compiling cast v0.2.3
   Compiling quote v1.0.9
    Checking futures v0.3.15
   Compiling darling_core v0.10.2
   Compiling pin-project-internal v1.0.7
   Compiling darling_macro v0.10.2
   Compiling darling v0.10.2
   Compiling embassy-macros v0.1.0 (/home/rukai2/Projects/Crates/embassy/embassy-macros)
    Checking pin-project v1.0.7
    Finished dev [unoptimized + debuginfo] target(s) in 5.67s
rukai@memes embassy$ cargo dylint --path /home/rukai2/Projects/Crates/deny_core_fmt/deny_core_fmt_dylint/target/release/libdeny_core_fmt@nightly-2021-07-01-x86_64-unknown-linux-gnu.so
If you don't see a backtrace below, it could be because `cargo-dylint` wasn't built with a nightly compiler.
Error: expected more input

Stack backtrace:
   0: dylint::driver_builder::get
   1: dylint::run_with_name_toolchain_map
   2: dylint::run
   3: cargo_dylint::main
   4: std::sys_common::backtrace::__rust_begin_short_backtrace
   5: std::rt::lang_start::{{closure}}
   6: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/core/src/ops/function.rs:259:13
   7: std::panicking::try::do_call
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/panicking.rs:401:40
   8: std::panicking::try
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/panicking.rs:365:19
   9: std::panic::catch_unwind
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/panic.rs:434:14
  10: std::rt::lang_start_internal::{{closure}}
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/rt.rs:45:48
  11: std::panicking::try::do_call
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/panicking.rs:401:40
  12: std::panicking::try
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/panicking.rs:365:19
  13: std::panic::catch_unwind
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/panic.rs:434:14
  14: std::rt::lang_start_internal
             at /rustc/240ff4c4a0d0936c9eeb783fa9ff5c0507a6ffb4/library/std/src/rt.rs:45:20
  15: main
  16: __libc_start_main
  17: _start
```


Running on a trival project I dont get that error but the lint never fails when it clearly should.

**Comment by smoelius** [comment]
Your output suggests that the driver was built correctly. So there should be an executable at this location:
```
~/.dylint_drivers/nightly-2021-07-01-x86_64-unknown-linux-gnu/dylint-driver
```
Could you run it (with no arguments) and tell me what it outputs?

**Comment by rukai** [comment]
```
rukai@memes ~$ ~/.dylint_drivers/nightly-2021-07-01-x86_64-unknown-linux-gnu/dylint-driver
nightly-2021-07-01-x86_64-unknown-linux-gnu 1.0.0
```

**Comment by smoelius** [comment]
So I think I 90% understand what's going on.

The root cause appears to be this: https://github.com/trailofbits/dylint#limitations

Dylint is trying to build `embassy` using a particular toolchain. But `embassy` has a `rust-toolchain` file that insists on another toolchain. When the driver is run, the presence of that `rust-toolchain` file causes the driver to link against some version of a library the driver doesn't expect (this is the part I don't completely understand). So the driver fails to execute, its output is empty, and the version parser reports that it expects more input.

The only workarounds I can see are to remove `embassy`'s `rust-toolchain` file, or edit it.

Still, there's definitely a bug here: Dylint should notice and report that the driver failed, rather than blindly accept empty output.

Also, from our earlier conversation, the documentation should indicate that `cargo-dylint` cannot be installed with `--debug`.

Regarding inspecting `cargo-dylint`'s path: the problem is that the installation path [can change](https://doc.rust-lang.org/cargo/commands/cargo-install.html#description). So I am afraid that that solution could make Dylint unusable for certain users, or could be fragile. Thanks for the suggestion, though.

I'll the get the error handling and the documentation fixed. Thanks a lot for exposing these problems! Please let me know if I missed anything.

**Comment by PatchMixolydic** [comment]
I seem to have run into a similar issue while trying to update [my personal lint set](https://github.com/PatchMixolydic/patchmixolint). This time, it manifested as a runtime link error:
```
$ cargo dylint target/debug/libpatchmixolint@nightly-x86_64-unknown-linux-gnu.so 
Error: command failed: "/home/patchmixolydic/.dylint_drivers/nightly-x86_64-unknown-linux-gnu/dylint-driver" "-V"
stdout: ""
stderr: "/home/patchmixolydic/.dylint_drivers/nightly-x86_64-unknown-linux-gnu/dylint-driver: error while loading shared libraries: librustc_driver-e924046ea34a2c7b.so: cannot open shared object file: No such file or directory\n"
```
Deleting `~/.dylint-driver` fixed the issue once again. I'm not entirely sure if my use case falls under the limitations listed in the README since I was trying to lint the lint library itself.

`cargo dylint --version`: `cargo-dylint 1.0.9`
`rustc -Vv`:
```
rustc 1.59.0-nightly (0b6f079e4 2021-12-07)
binary: rustc
commit-hash: 0b6f079e4987ded15c13a15b734e7cfb8176839f
commit-date: 2021-12-07
host: x86_64-unknown-linux-gnu
release: 1.59.0-nightly
LLVM version: 13.0.0
```

**Comment by smoelius** [comment]
Well, that's not good.

Would it happen that the `rust-toolchain` `channel` is `nightly` as opposed to `nightly-yyyy-mm-dd`?

I'm asking based on the `command failed` message, where I'm used to seeing `libname@nightly-yyyy-mm-dd-x86_64-unknown-linux-gnu.so`.

**Comment by PatchMixolydic** [comment]
[Indeed it is](https://github.com/PatchMixolydic/patchmixolint/blob/ac8a3b3fa8a5370cd66f7a39d29a84015ea5e8cc/rust-toolchain). Could this be preventing `cargo dylint` from realizing that the driver needs to be updated?

**Comment by smoelius** [comment]
Yeah, `cargo dylint` assumes that a driver can be built "once and for all" for any given toolchain. But obviously that assumption is naive.

Thanks a lot for raising this. This is a bug.
